File: trunk/项目开发进度2009年2月9日/SeeMapCell/MapXtreme/6.6/Samples/WebControlsSourceCode/vb/MapControl/MapControl.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Web
Imports System.Web.SessionState
Imports System.Web.UI
Imports System.Web.UI.WebControls
'/	<summary>Provides export formats for the web environment.</summary>
'/	<remarks>This enumeration was created to support only those export formats in
'/ the MapExportFormat that work in the web environment.</remarks>
Public Enum WebExportFormat
    '/	<summary>Exports Windows Bitmap formats.</summary>
    Bmp
    '/	<summary>Exports Graphics Interchange Format (GIF) formats.</summary>
    Gif
    '/	<summary>Exports Joint Photographic Experts Group (JPEG) formats.</summary>
    Jpeg
End Enum

<Designer(GetType(MapControlDesigner)), ToolboxData("<{0}:MapControl runat=server></{0}:MapControl>")> _
Public Class MapControl
    Inherits WebControl
    <PersistenceMode(PersistenceMode.Attribute), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)> _
    Public Overrides Property Width() As Unit
        Get
            If MyBase.Width.IsEmpty Then
                MyBase.Width = New Unit(300)
            End If
            Return MyBase.Width
        End Get
        Set(ByVal value As Unit)
            MyBase.Width = value
        End Set
    End Property

    <PersistenceMode(PersistenceMode.Attribute), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)> _
    Public Overrides Property Height() As Unit
        Get
            If MyBase.Height.IsEmpty Then
                MyBase.Height = New Unit(300)
            End If
            Return MyBase.Height
        End Get
        Set(ByVal value As Unit)
            MyBase.Height = value
        End Set
    End Property
    Private _mapAlias As String
    <PersistenceMode(PersistenceMode.Attribute), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)> _
    Public Property MapAlias() As String
        Get
            Return Me._mapAlias
        End Get
        Set(ByVal value As String)
            Me._mapAlias = value
        End Set
    End Property
    Private _imgUseMap As String
    '/ <summary>
    '/ UseMap attribute for the IMG tag which is rendered with this mapcontrol
    '/ </summary>
    '/ <remarks>Sometimes users want to usemaps with the map image. The value of this property will be added as extra UseMap attribute to 
    '/ the IMG tag rendered with this mapcontrol.</remarks>
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), PersistenceMode(PersistenceMode.Attribute)> _
    Public Property IMGUseMap() As String
        Get
            Return _imgUseMap
        End Get
        Set(ByVal Value As String)
            _imgUseMap = Value
        End Set
    End Property

    '/	<summary>Gets or sets the format used to export a map for rendering.</summary>
    '/	<value>Gets or sets the format used for exporting.</value>
    '/	<remarks>This property is used internally to export a map to be used for
    '/ rendering.</remarks>
    <Browsable(True), DefaultValue("GIF"), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), PersistenceMode(PersistenceMode.Attribute)> _
    Public Property ExportFormat() As WebExportFormat
        Get
            Dim format As WebExportFormat
            If ViewState("ExportFormat") Is Nothing Then
                format = WebExportFormat.Gif
            Else
                format = CType(ViewState("ExportFormat"), WebExportFormat)
            End If
            Return format
        End Get
        Set(ByVal Value As WebExportFormat)
            ViewState("ExportFormat") = value
        End Set
    End Property

    '/ <summary> 
    '/ Render the control by setting the source of the image to a url containing all information to get the right map
    '/ </summary>
    '/ <remarks>It renders the control html by writing html IMG tag and it's source is set to the url. Using javascript it also sets the mapalias property for the
    '/ IMG tag, so that other's can get to it after the image has been loaded, using OnLoad event of the IMG tag.
    '/ </remarks>
    '/ <param name="output"> The HTML writer object to write html to </param>
    Protected Overrides Sub RenderContents(ByVal output As HtmlTextWriter)
        ' Get the model out of ASP.NET and get the stream
        Dim model As MapControlModel = MapControlModel.GetModelFromSession()
        If model Is Nothing Then
            model = MapControlModel.SetDefaultModelInSession()
        End If
        Dim ms As System.IO.MemoryStream = Nothing
        Try
            ms = model.GetMap(MapAlias, CType(Width.Value, Integer), CType(Height.Value, Integer), ExportFormat.ToString())
        Catch ex As Exception
            output.WriteLine(ex.Message)
            output.WriteLine("<br>")
            output.WriteLine(L10NUtils.Resources.GetString("MapNotFoundErrorString"))
            HttpContext.Current.Server.ClearError()
            Return
        End Try

        ' Insert the image stream to Cache with key imageid and timeout in 2 mintues.
        Dim imageid As String = ImageHelper.GetUniqueID()
        ImageHelper.SetImageToCache(imageid, ms, 2)
        Dim url As String = ImageHelper.GetImageURL(imageid, ExportFormat.ToString())

        '//Write the IMG tag and set the mapalias.
        output.AddAttribute(HtmlTextWriterAttribute.Width, Me.Width.ToString)
        output.AddAttribute(HtmlTextWriterAttribute.Height, Me.Height.ToString)
        If (Not IMGUseMap = Nothing) Then
            If (IMGUseMap.Length > 0) Then
                output.AddAttribute("USEMAP", IMGUseMap)
            End If
        End If
        If (Not CssClass = Nothing) Then
            output.AddAttribute(HtmlTextWriterAttribute.Class, CssClass)
        End If

        output.AddAttribute(HtmlTextWriterAttribute.Alt, "")
        output.AddAttribute(HtmlTextWriterAttribute.Src, url)
        output.AddAttribute(HtmlTextWriterAttribute.Id, (Me.UniqueID & "_Image"))
        output.RenderBeginTag(HtmlTextWriterTag.Img)
        output.RenderEndTag()
        output.Flush()

        'Now set the alias and format manually by getting the image element.
        output.AddAttribute("language", "javascript")
        output.AddAttribute(HtmlTextWriterAttribute.Type, "text/javascript")
        output.RenderBeginTag(HtmlTextWriterTag.Script)
        output.WriteLine(String.Format("var {0}Me = document.getElementById('{0}_Image');", UniqueID))
        output.WriteLine(String.Format("{0}Me.mapAlias= '{1}';", UniqueID, MapAlias))
        output.WriteLine(String.Format("{0}Me.exportFormat= '{1}';", UniqueID, ExportFormat))
        output.RenderEndTag()

    End Sub
End Class

File: MapXtreme/6.6/Samples/WebControlsSourceCode/cs/MapControl/MapControl.cs
using System;
using System.ComponentModel;
using System.Drawing.Design;
using System.IO;
using System.Web;
using System.Web.UI;
using System.Web.UI.Design;
using System.Web.UI.WebControls;
using System.Windows.Forms.Design;
using System.Xml;
using System.Xml.XPath;
using System.Xml.Xsl;

namespace MapInfo.WebControls
{
	///	<summary>Provides export formats for the web environment.</summary>
	///	<remarks>This enumeration was created to support only those export formats in
	/// the MapExportFormat that work in the web environment.</remarks>
	public enum WebExportFormat { 
		///	<summary>Exports Windows Bitmap format.</summary>
		Bmp,
		///	<summary>Exports Graphics Interchange Format (GIF) format.</summary>
		Gif,
		///	<summary>Exports Joint Photographic Experts Group (JPEG) format.</summary>
		Jpeg
	}

	///	<summary>Contains an instance of a map that is displayed.</summary>
	///	<remarks><para>The MapControl contains an instance of a Map object. This Map is obtained
	/// from a MapFactory by using the MapAlias property it is pointing to. The
	/// map is drawn by exporting the map into an image and this image is then used in the
	/// IMG HTML tag.  If the MapAlias property is not specified or is invalid, the first map from the
	/// MapFactory is chosen.</para>
	/// <para>The MapControl and other controls in this assembly follow the MVC (Model-View-Controller) pattern.
	/// The Control code draws the HTML and javascript, the trip to the server to perform the actual operation is done through the 
	/// httphandler. This way the map gets updated without re-drawing the entire page, allowing you to get visually appealing updates.
	/// The model talks to the MapXtreme core engine to perform the job.</para>
	/// <para>In order to draw a correct map, follow these steps:</para>
	/// <nl><item>Set up the workspace. It may contain single or multiple maps. When multiple maps are used, the recommended approach is to put all maps in a single workspace and draw a specific map through the MapAlias,
	/// instead of loading and unloading different workspaces.</item>
	/// <item>Choose the initial MapAlias for the map or leave it blank. If left blank, the first map in the 
	/// MapFactory will be chosen.</item></nl>
	/// <para>Once these steps are complete, the MapControl contacts the MapController with the appropriate information in the URL, 
	/// gets the exported image, and then the image is drawn.</para>
	///  </remarks>
	[
	ToolboxData("<{0}:MapControl runat=server></{0}:MapControl>"),
	Designer(typeof(MapControlDesigner))
	]
	public class MapControl : WebControl {
		/// <summary>
		/// Width of the map. 
		/// </summary>
		/// <remarks>The width is passed to the controller which tells the model to use the width for the exported map.</remarks>
		[
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute),
		]
		public override Unit Width {
			get {
				if (base.Width.IsEmpty) {base.Width = new Unit(300);}
				return base.Width;
			}
			set {
				base.Width = value;
			}
		}

		/// <summary>
		/// Height of the map. 
		/// </summary>
		/// <remarks>The height is passed to the controller which tells the model to use the height for the exported map.</remarks>
		[
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute),
		]
		public override Unit Height {
			get {
				if (base.Height.IsEmpty) {base.Height = new Unit(300);}
				return base.Height;
			}
			set {
				base.Height = value;
			}
		}

		private string _mapAlias;
		/// <summary>
		/// MapAlias to get a specific map. 
		/// </summary>
		/// <remarks>If this is not provided or is invalid, then the first map in the MapFactory is chosen.</remarks>
		[
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute)
		]
		public string MapAlias {
			get {
				return _mapAlias;
			}
			set {
				_mapAlias = value;
			}
		}

		private string _imgUseMap;
		/// <summary>
		/// UseMap attribute for the IMG tag which is rendered with the MapControl
		/// </summary>
		/// <remarks>Sometimes users want to usemaps with the map image. The value of this property will be added as an extra UseMap attribute to 
		/// the IMG tag rendered with this mapcontrol.</remarks>
		[
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute)
		]
		public string IMGUseMap {
			get {
				return  _imgUseMap;
		}
			set {
				_imgUseMap= value;
			}
		}

		///	<summary>Gets or sets the format used to export a map for rendering.</summary>
		///	<value>Gets or sets the format used for exporting.</value>
		///	<remarks>This property is used internally to export a map to be used for
		/// rendering.</remarks>
		[
		Browsable(true),
		DefaultValue("GIF"),
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute),
		] 
		public WebExportFormat ExportFormat {
			get {
				WebExportFormat format; 
				if(ViewState["ExportFormat"] == null)
					format = WebExportFormat.Gif;
				else
					format = (WebExportFormat)ViewState["ExportFormat"];
				return format;
			}

			set {
				ViewState["ExportFormat"] = value;
			}
		}

		/// <summary> 
		/// Renders the control by setting the source of the image to a URL containing all information to get the right map.
		/// </summary>
		/// <remarks>This method renders the controls HTML by writing an HTML IMG tag setting its source to the URL. 
		/// </remarks>
		/// <param name="output">The HTML writer object to write HTML to.</param>
		protected override void RenderContents(HtmlTextWriter output) 
		{
			// Get the model out of ASP.NET and get the stream
			MapControlModel model = MapControlModel.GetModelFromSession();
			if (model == null) model = MapControlModel.SetDefaultModelInSession();
			MemoryStream ms = null;
			try{
				ms = model.GetMap(MapAlias, (int)Width.Value, (int)Height.Value, ExportFormat.ToString());
			} catch(Exception ex) {
				output.WriteLine(ex.Message);
				output.WriteLine("<br>");
				output.WriteLine(L10NUtils.Resources.GetString("MapNotFoundErrorString"));
				HttpContext.Current.Server.ClearError();
				return;
 				}

			// Insert the image stream to Cache with key imageid and timeout in 2 mintues.
			string imageid = ImageHelper.GetUniqueID();
			ImageHelper.SetImageToCache(imageid, ms, 2);
			string url = ImageHelper.GetImageURL(imageid, ExportFormat.ToString());

			//Write the IMG tag and set the mapalias.
			output.AddAttribute(HtmlTextWriterAttribute.Width, Width.ToString());
			output.AddAttribute(HtmlTextWriterAttribute.Height, Height.ToString());
//			output.AddAttribute("exportFormat", ExportFormat.ToString());
	//		output.AddAttribute("mapAlias", MapAlias);
			if (IMGUseMap != null) {
				if (IMGUseMap.Length > 0) output.AddAttribute("USEMAP", IMGUseMap);
			}
			if (CssClass != null) output.AddAttribute(HtmlTextWriterAttribute.Class, CssClass);
	        output.AddAttribute(HtmlTextWriterAttribute.Alt, "");
			output.AddAttribute(HtmlTextWriterAttribute.Src, url);
			output.AddAttribute(HtmlTextWriterAttribute.Id, UniqueID + "_Image");
			output.RenderBeginTag(HtmlTextWriterTag.Img);
			output.RenderEndTag();
			output.Flush();

	        // Now set the alias and format manually by getting the image element.
			output.AddAttribute("language", "javascript");
			output.AddAttribute(HtmlTextWriterAttribute.Type, "text/javascript");
			output.RenderBeginTag(HtmlTextWriterTag.Script);
			output.WriteLine(string.Format("var {0}Me = document.getElementById('{0}_Image');", UniqueID));
			output.WriteLine(string.Format("{0}Me.mapAlias= '{1}';", UniqueID, MapAlias));
			output.WriteLine(string.Format("{0}Me.exportFormat= '{1}';", UniqueID, ExportFormat));
			output.RenderEndTag();
		}
	}
}
